Feature/scrutinizer optimizations

p4u opened the pull request:
see commit message

p4u committed:
vochaintest: add new flag --checkNulliriers

Checking all nullifiers one by one is very slow and costly.
We can just relay on the vote envelope height and the final
results in order to verify the votes.

The flag is false by default.

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
scrutinizer refactor and optimizations

This refactor is product of a deep benchmarking performed during some days
with a huge load of votes (up to 10 Million). Some weakness have been
found so they are addressed here.

Use type [][]*big.Int instead of protobuf bytes for storing the vote
results since it's more efficient when performing operations (adding
votes).

Instead of make the scrutinizer count always all live votes, count only
live votes after fastsync is finished. This increases the bootstrap time
for the Vochain.

Make the live votes counting async, so it does not have a huge impact on
the consensus block time.

Handle the computation of final results for unencrypted processes the
same way we compute the encrypted ones. The cost of counting all votes
at once is much lower than computing one-by-one as they are introduced in the
blockchain. So when final results computation is executed, the live
votes result will be override and the goroutines that already need to
compute live votes will be halted.

When counting live votes do not make 1 db transaction per vote but
aggregate them by processId and make a single transaction per block and
processId.

Show the last votes per minute (vxm) on the [vochain info] log message.

Signed-off-by: p4u <pau@dabax.net>

mvdan commented on chain/ethevents/events.go:
```diff
@@ -150,60 +150,66 @@ func NewEthEvents(
 	default:
-		log.Infof("invalid process %x status %s for setting the results, skipping", results.ProcessId, vocProcessData.Status)
+		log.Infof("invalid process %x status %s for setting the results, skipping",
+			results.ProcessID, vocProcessData.Status.String())
```
no need to call String if the verb is already %s

mvdan commented on client/api.go:
```diff
@@ -404,27 +403,29 @@ func (c *Client) TestSendVotes(
-			if nullHex == "registered" {
-				registered++
-				continue
+	if checkNullifiers {
```
invert and early return to unindent the rest

mvdan commented on cmd/vochaintest/vochaintest.go:
```diff
@@ -45,6 +45,7 @@ func main() {
 	parallelCons := flag.Int("parallelCons", 1, "parallel API connections")
 	procDuration := flag.Int("processDuration", 500, "voting process duration in blocks")
 	doubleVote := flag.Bool("doubleVote", true, "send every vote twice")
+	checkNullifiers := flag.Bool("checkNullifiers", false, "check all votes are correct one by one (slow)")
```
So, turning this on is slower but... more secure? Or more reliable? I'm not sure if I get why anyone would ever want to turn it on.

If there's no reason to turn it on, I'm inclined to just not have this feature at all. And if the reason to turn it on is better security/reliability, I wonder why it's not on always.

p4u replied:
This is a testing tool. I think the full check of nullifiers could be useful depending on the purpose of the testing. I will keep it available, it's not a big chunk of code.

mvdan commented on vochain/state.go:
```diff
@@ -45,17 +45,25 @@ var PrefixDBCacheSize = 0
 // happening in parallel.
+//
+// If Commit() returns an error and isFatal set tu true, it is considered a consensus
+// failure and the blockchain will halt.
```
typo: "tu true"

I also think that just an `error` return would be simpler and clearer here. You could then use a special error type to specify fatal errors. For example:

```
type ErrHaltVochain struct {
    reason error
}

func (e ErrHaltVochain) Error() string { return fmt.Sprintf("halting vochain: %v", e.reason) }
func (e ErrHaltVochain) Unwrap() error { return e.reason }
```

Then, instead of:

```
err, fatal := l.Commit(h.Height)
if err != nil {
	if fatal {
		panic(err)
	}
	log.Warn(err)
}
```

you could do

```
err := l.Commit(h.Height)
if err != nil {
	if _, fatal := err.(ErrHaltVochain); fatal {
		panic(err)
	}
	log.Warn(err)
}
```

Or, even simpler, just panic directly in Commit if there's a fatal error. Having the caller panic is really the same, but more complex.

p4u replied:
Great solution. Thanks.

mvdan commented on vochain/state.go:
```diff
@@ -100,19 +110,24 @@ func NewState(dataDir string) (*State, error) {
 	if err := state.Store.Init(dataDir, "disk"); err != nil {
 		return err
 	}
+	timer := time.Now()
```
this is not a timer, so maybe call it `startTime`

mvdan commented on vochain/state.go:
```diff
@@ -122,6 +137,7 @@ func initStore(dataDir string, state *State) error {
 	if err := state.Store.AddTree(VoteTree); err != nil {
 		return err
 	}
+	log.Infof("state tree load took %d ms", time.Since(timer).Milliseconds())
```
I'd just use `%s` with `time.Since(startTime)`, it will already do the right thing by printing e.g. `200ms`. It will also behave better if the duration is over 1s.

mvdan commented on vochain/scrutinizer/db.go:
```diff
@@ -47,9 +47,25 @@ type Entity struct {
+}
+
+func (r *Results) String() string {
+	results := ""
```
use a `strings.Builder`; otherwise you're allocating and copying the entire string at ever `+=` statement. Since there could be thousands of votes, this could be *very* slow, and indeed quadratic.

mvdan commented on vochain/scrutinizer/scrutinizer.go:
```diff
@@ -34,16 +38,19 @@ func (s *Scrutinizer) AddEventListener(l EventListener) {
 	// eventListeners is the list of external callbacks that will be executed by the scrutinizer
 	eventListeners []EventListener
 	db             *badgerhold.Store
+	addVoteLock    sync.RWMutex
```
please document what this mutex is actually guarding.

mvdan replied:
For example, right now I don't get it - the DB is already atomic and thread-safe, so having a mutex around inserts/gets of votes seems entirely unnecessary. And if this is preventing data races, then we probably have a bug somewhere.

p4u replied:
That's for avoiding the "Transaction Conflict" error which happens quite often. That's a non critical error and the badger developers recommend to try it again until the db transaction works. But I preferred to try to avoid directly the put/get conflict with a Mutex. And yes, with that Mutex the iteration over the "Transaction Conflict" error might be removed but I let it there for security since I'm still not 100% sure all conflicting calls are mutex.

mvdan replied:
If retrying works, then why have the mutex? I don't know, having a mutex around the entire DB feels hacky. If we have good tests and they continuously pass without the mutex and the error retry, then I'd say that's good.

p4u replied:
Well, to me it looks that repeating a DB transaction N times until it success is kind of hacky. Using a Mutex on this scope makes sense and will release load on the KV context. 

In addition I added a `maxRetries=1000` mechanism which is kind of arbitrary (I tested different values until there were no more transactions failing). This value works in my local computer but a slower one could need more maxRetries. I don't like to block the goroutine forever if there is an issue on the KV (that won't accept new transactions) I prefer instead to try N times and raise an Error. So this is why I'm mixing both approaches: Mutex + Repeat Transaction N times

I'll add a to-do for removing it in the future, with proper research and testing involved. But for now I prefer to make it safe.

mvdan commented on vochain/scrutinizer/scrutinizer.go:
```diff
@@ -56,52 +63,103 @@ func NewScrutinizer(dbPath string, state *vochain.State) (*Scrutinizer, error) {
 }
 
+func (s *Scrutinizer) AfterSyncBootstrap() {
+	// Some grace time to avoid false positive on IsSynchronizing()
```
I don't get this...

p4u replied:
During the first seconds/milliseconds of the Vochain startup, Tendermint might report that the chain is not synchronizing since it still does not have any peer and do not know the actual size of the blockchain. If AfterSyncBootStrap is executed on this specific moment, the Wait loop would pass.

mvdan replied:
Gotcha, maybe summarize that here?

mvdan commented on vochain/scrutinizer/scrutinizer.go:
```diff
@@ -56,52 +63,103 @@ func NewScrutinizer(dbPath string, state *vochain.State) (*Scrutinizer, error) {
+	time.Sleep(time.Second * 10)
+	for s.VochainState.IsSynchronizing() {
+		time.Sleep(time.Second * 2)
+	}
```
is there no way to block until the state is synchronized? that would prevent the polling/sleeping, and supposedly also fix the 10s sleep at the beginning too.

For example, have one channel that is closed when the state is fully synced. Then, here we just need to `<-s.VochainState.Synced()` or something like that. When the channel is closed, that receive will unblock.

p4u replied:
With the channel we would still have the same problem.

Maye a more safe and dynamic approach would be something like:
```
        syncSignal := 5
	for {
	is !s.VochainState.IsSynchronizing() { syncSignal-- }
	if syncSignal == 0 { break }
        time.Sleep(time.Second * 1)
	}
```

mvdan commented on vochain/scrutinizer/scrutinizer.go:
```diff
@@ -56,52 +63,103 @@ func NewScrutinizer(dbPath string, state *vochain.State) (*Scrutinizer, error) {
 		if err := s.updateProcess(p); err != nil {
-			log.Warnf("commit: cannot update process %x: %v", p, err)
+			log.Errorf("commit: cannot update process %x: %v", p, err)
+			continue
```
this loop does nothing else, so I guess you can remove the continue

mvdan commented on vochain/scrutinizer/scrutinizer.go:
```diff
@@ -56,52 +63,103 @@ func NewScrutinizer(dbPath string, state *vochain.State) (*Scrutinizer, error) {
+		}
+		if nvotes > 0 {
+			log.Infof("added %d live votes on block %d, took %d ms",
+				nvotes, height, time.Since(timer).Milliseconds())
```
same comments about `timer` and `%s` like before :)

mvdan commented on vochain/scrutinizer/scrutinizer.go:
```diff
@@ -56,52 +63,103 @@ func NewScrutinizer(dbPath string, state *vochain.State) (*Scrutinizer, error) {
+			results := &Results{Weight: new(big.Int).SetUint64(0)}
+			for _, v := range votes {
+				if err := s.addLiveVote(v, results); err != nil {
+					log.Warnf("vote cannot be added: %v", err)
```
shouldn't this be Errorf?

p4u replied:
Warning is fine since the user might provide a wrong Vote package and that should be considered as a "null vote", it's not an error and probably in the future (when we are sure everything works as expected) we can even remove the warning

mvdan commented on vochain/scrutinizer/scrutinizer.go:
```diff
@@ -115,7 +173,11 @@ func (s *Scrutinizer) OnProcess(pid, eid []byte, censusRoot, censusURI string) {
 func (s *Scrutinizer) OnVote(v *models.Vote) {
-	s.votePool = append(s.votePool, v)
+	if _, ok := s.votePool[string(v.ProcessId)]; ok {
+		s.votePool[string(v.ProcessId)] = append(s.votePool[string(v.ProcessId)], v)
```
I think you can just have this line. Because if the map key doesn't exist, you'll do `append` with the first argument being nil, which will work just fine.

mvdan commented on vochain/scrutinizer/vote.go:
```diff
@@ -163,80 +135,125 @@ func (s *Scrutinizer) addLiveVote(envelope *models.Vote) error {
+		}
+		update.Weight.Add(update.Weight, results.Weight)
+
+		if len(results.Votes) > 0 {
```
invert and early return to unindent

mvdan commented on vochain/scrutinizer/vote.go:
```diff
@@ -163,80 +135,125 @@ func (s *Scrutinizer) addLiveVote(envelope *models.Vote) error {
+		if err == nil {
+			break
+		}
+		if strings.Contains(err.Error(), "Transaction Conflict") {
```
use badger.ErrConflict instead: https://pkg.go.dev/github.com/dgraph-io/badger/v2#pkg-variables

If the error is wrapped, then use `errors.Is(err, badger.ErrConflict)`.

p4u replied:
Mmm, I did not add this error cause I did not want to import a specific KV implementation. So badgerhold can switch from badger v2 to v3 without any change in our side. Do you really think it is worth it? (importing badger v2 just for handling this error).

mvdan replied:
Right, I wasn't thinking about v2 vs v3. If we have good tests which cover the conflict case (e.g. with many reads/writes), then I imagine the tests would catch if they bump to v3 but our error points at v2.

Alternatively, add a short comment to explain the strings.Contains? Most times it's a mistake, so it would be good to clarify why it's wanted here.

mvdan commented on vochain/scrutinizer/vote.go:
```diff
@@ -163,80 +135,125 @@ func (s *Scrutinizer) addLiveVote(envelope *models.Vote) error {
+	}
+	if err != nil {
+		log.Debugf("saved %d votes with total weight of %s on process %x", len(results.Votes),
+			results.Weight.String(), pid)
```
no need to call String with %s

mvdan reviewed (commented):
I can't really approve because I fail to properly understand half of the changes :) If you have the time and want me to look more carefully, maybe we can do a 30m call and I'll ask questions while reading the changes.

p4u committed:
apply review comments

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
scrutinizer: finish the recovery mechanism

Since live votes are not computed during Fast Sync and we don't know
which envelopes have been counted when starting the blockchain, we need
to execute a bootstrap live results recovery.

Apply also some review comments.

Signed-off-by: p4u <pau@dabax.net>

mvdan committed:
don't re-initialize the logger in each test

The logger is global, and already initialized by default as of
6d68b9cf96.

These tests were re-initializing the logger while others were still
logging stuff, causing data races:

	WARNING: DATA RACE
	Read at 0x0000024de2c0 by goroutine 81:
	  go.vocdoni.io/dvote/log.Infof()
	      /src/dvote/log/log.go:157 +0xdb
	  go.vocdoni.io/dvote/vochain/scrutinizer.(*Scrutinizer).afterSyncBootstrap()
	      /src/dvote/vochain/scrutinizer/scrutinizer.go:98 +0xcb

	Previous write at 0x0000024de2c0 by goroutine 261:
	  go.vocdoni.io/dvote/log.Init()
	      /src/dvote/log/log.go:41 +0x447
	  go.vocdoni.io/dvote/vochain/scrutinizer.TestResults()
	      /src/dvote/vochain/scrutinizer/scrutinizer_test.go:219 +0x72
	  testing.tRunner()
	      /tip/src/testing/testing.go:1235 +0x22d

p4u committed:
disable Tendermint tx indexer and fix scrutinizer race

Since we don't use the tendermint transaction indexer it is worth to
disable it. Its impact on performance and disk size is very important.

On scrutinizer do not execute AfetSyncBootstrap on the constructor, so
it will only be enabled if the developer wants. For testing its not
required and its the cause for some data races.

Signed-off-by: p4u <pau@dabax.net>
